app/presenters/compute_resource_presenter.rb

```diff
@@ -0,0 +1,93 @@
+class ComputeResourcePresenter
+  def self.for_cr(compute_resource, view)
+    klass = "#{compute_resource.provider.camelcase}Presenter".safe_constantize
+    Rails.logger.warn("Do not know how to present #{compute_resource.provider}") unless klass
+    klass ||= self
+    klass.new(compute_resource, view)
+  end
+
+  attr_accessor :model, :view
+  alias_method :compute_resource, :model
+
+  def initialize(compute_resource, view)
+    @model, @view = compute_resource, view
+  end
+
+  # TODO: preparation of the cache should be private in the future
+  def prepare_host_cache!(vms)
+    @hosts = Host.for_vm(compute_resource, vms).group_by(:uuid)
+  end
+
+  def prepare_host_existance_cache!(vms)
+    @existing_hosts = Host.for_vm(compute_resource, vms).pluck(:uuid)
```

Comment:
Based on my comment for `prepare_host_cache!`, we could reuse the same `hosts_cache` hash to calculate existence:

``` ruby
def existing_hosts
  hosts.keys
end
```

This way we are not hitting the DB again.